feat(compiler-cli): report error if undecorated class with Angular features is discovered

devversion opened the pull request:
* See individual commits.

devversion committed:
feat(core): undecorated-classes-with-decorated-fields migration should handle classes with lifecycle hooks

As of v10, undecorated classes using Angular features are no longer
supported. In v10, we plan on removing the undecorated classes
compatibility code in ngtsc. This means that old patterns for
undecorated classes will result in compilation errors.

We had a migration for this in v9 already, but it looks like
the migration does not handle cases where classes uses lifecycle
hooks. This is handled in the ngtsc compatibility code, and we
should handle it similarly in migrations too.

This has not been outlined in the migration plan initially,
but an appendix has been added for v10 to the plan document.

https://hackmd.io/vuQfavzfRG6KUCtU7oK_EA?both.

Note: The migration is unable to determine whether a given undecorated
class that only defines `ngOnDestroy` is a directive or an actual
service. This means that in some cases the migration cannot do
more than adding a TODO and printing an failure.

Certainly there are more ways to determine the type of such classes,
but it would involve metadata and NgModule analysis. This is out of
scope for this migration.

devversion committed:
feat(compiler-cli): report error if undecorated class with Angular features is discovered

Previously in v9, we deprecated the pattern of undecorated base classes
that rely on Angular features. We ran a migration for this in version 9
and will run the same on in version 10 again.

To ensure that projects do not regress and start using the unsupported
pattern again, we report an error in ngtsc if such undecorated classes
are discovered.

We keep the compatibility code enabled in ngcc so that libraries
can be still be consumed, even if they have not been migrated yet.

Resolves FW-2130.

devversion committed:
refactor: enable ng update migrations for v10

Enables the `ng update` migrations for v10. Status for individual
migrations:

**undecorated-classes-with-di**.

This migration dealt exlusively with inherited constructors and
cases where a derived component was undecorated. In those cases,
the migration added `@Directive()` or copied the inherited decorator
to the derived class.

We don't need to run this migration again because ngtsc throws if
constructor is inherited from an undecorated class. Also ngtsc will
throw if a NgModule references an undecorated class in the declarations.

***undecorated-classes-with-decorated-fields***

This migration exclusively deals with undecorated classes that use
Angular features but are not decorated. Angular features include
the use of lifecycle hooks or class fields with Angular decorators,
such as `@Input()`.

We want to re-run this migration in v10 as we will disable the
compatibility code in ngtsc that detects such undecorated classes
as `@Directive`.

**module-with-providers**:

This migration adds an explicit generic type to `ModuleWithProviders`.
As of v10, the generic type is required, so we need to re-run the
migration again.

**renderer-to-renderer2**:

We don't need to re-run that migration again as the
renderer has been already removed in v9.

**missing-injectable**:

This migration is exclusively concerned with undecorated
providers referenced in an `NgModule`. We should re-run
that migration again as we don't have proper backsliding
prevention for this yet. We can consider adding an error
in ngtsc for v10, or v11. In either way, we should re-run
the migration.

**dynamic-queries**:

We ran this one in v9 to reduce code complexity in projects. Instead
of explicitly passing `static: false`, not passing any object literal
has the same semantics. We don't need to re-run the migration again
since there is no good way to prevent backsliding and we cannot always
run this migration for future versions (as some apps might actually
intentionally use the explicit `static: false` option).

devversion committed:
refactor: fix undecorated classes with angular features in repo

A few instances of undecorated classes with Angular features
have been discovered in the framework repo. This commit fixes
those.

devversion commented:
Note: The components repo unit test job currently fails as a new compiler diagnostic has been introduced in this PR and errors would need to be solved in the components repo. 

Regardless of that failing job though, it would be good to receive feedback on the PR.

atscott reviewed (approved):
reviewed for: fw-router

devversion committed:
build(docs-infra): do not require abstract directives to declare a @ngModule

Abstract directives cannot be part of a `@NgModule`, but the AIO dgeni
setup currently enforces this. This commit updates the logic so that
abstract directives are skipped.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:
```diff
@@ -78,6 +78,14 @@ export function getDirectiveDiagnostics(
+    ts.Diagnostic {
+  return makeDiagnostic(
+      ErrorCode.UNDECORATED_CLASS_USING_ANGULAR_FEATURES, node.name,
+      `Class is using Angular features but is not decorated. Please add an explicit ` +
```
This is good for now, but we want to make this error message better before final.

1) the error should go on the feature in question, not the class identifier. So if you have an `@Input() field: ...` on the class, the error should go on the `@Input` decorator itself. Or the lifecycle method name, etc.

2) the error should include an extra pointer to the class name (a `relatedInfo` node) that suggests adding a decorator to the class itself. For `@Input` or any of the directive/component specific annotations, we should suggest adding `@Directive()`. For lifecycle hooks or DI annotations in the constructor, we should suggest adding any decorator.

devversion replied:
(1) 👍 
(2) 👍 I did something like that initially but left that out as it felt out-of-scope. I can re-use that. https://gist.github.com/devversion/d412695321ca2a37fc59d84b6255f733 (patch is reversed)

devversion replied:
@alxhub Regarding (2), you mention DI annotations. This is not something that is handled yet. It sounds like we should though? I assume we can do that as a follow-up too.

kara replied:
+1 to what Alex said. I'm good with that being a follow-up.

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -905,6 +907,23 @@ runInEachFileSystem(os => {
               `static ɵdir: i0.ɵɵDirectiveDefWithMeta<TestBase, never, never, { "input": "input"; }, {}, never>;`);
     });
 
+    it('should error if undecorated class with Angular features has been discovered', () => {
```
Add a test for the lifecycle case as well.

alxhub reviewed (approved):
LGTM with a few comments, for the compiler piece only.

AndrewKushnir reviewed (approved):
Reviewed for: fw-forms

kara reviewed (changes requested):
Main things look good. Typos and tests

petebacondarwin reviewed (approved):
Reviewed for: fw-upgrade, docs-infra

devversion commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -905,6 +907,44 @@ runInEachFileSystem(os => {
+                'Angular decorator.');
+      });
+
+      it('should error if directive lifecye hook has been discovered', () => {
```
lol. Not sure how that one got cut off. Good catch.

kara reviewed (approved):
LGTM

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers

devversion committed:
ci: update components-repo-unit-tests job commit

Updates the `components-repo-unit-tests` job to
https://github.com/angular/components/commit/d3a9ac67d2ef89ca9570c73af41111c796c135ca.

We need to update since we added a new diagnostic in ngtsc, and
the given commit in the components repo fixes failures caused by
the new diagnostic.

Note: This commit currently points to a PR as it's unlikely that
this fix lands soon, but we want to move forward. There is no
downside to doing that as the PR is based on top of the latest
components repo `master`.
